src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/StartupObjectValueProvider.cs

```diff
@@ -0,0 +1,89 @@
+﻿// Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. See the LICENSE.md file in the project root for more information.
+
+using Microsoft.VisualStudio.ProjectSystem.Properties;
+using Microsoft.VisualStudio.ProjectSystem.VS.WindowsForms;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Properties;
+
+[ExportInterceptingPropertyValueProvider(StartupObjectProperty, ExportInterceptingPropertyValueProviderFile.ProjectFile)]
+[AppliesTo(ProjectCapability.VisualBasic)]
+internal class StartupObjectValueProvider : InterceptingPropertyValueProviderBase
+{
+    private readonly IMyAppFileAccessor _myAppXmlFileAccessor;
+
+    internal const string ApplicationFrameworkProperty = "MyType";
+    private const string EnabledValue = "WindowsForms";
+    private const string DisabledValue = "WindowsFormsWithCustomSubMain";
+
+    internal const string UseWinFormsProperty = "UseWindowsForms";
+    internal const string StartupObjectProperty = "StartupObject";
+
+    [ImportingConstructor]
+    public StartupObjectValueProvider(IMyAppFileAccessor myAppXmlFileAccessor)
+    {
+        _myAppXmlFileAccessor = myAppXmlFileAccessor;
+    }
+
+    public override async Task<string?> OnSetPropertyValueAsync(string propertyName, string unevaluatedPropertyValue, IProjectProperties defaultProperties, IReadOnlyDictionary<string, string>? dimensionalConditions = null)
+    {
+        string isWindowsFormsProject = await defaultProperties.GetEvaluatedPropertyValueAsync(UseWinFormsProperty);
+
+        if (isWindowsFormsProject == "true")
+        {
+            string applicationFrameworkValue = await defaultProperties.GetEvaluatedPropertyValueAsync(ApplicationFrameworkProperty);
+
+            if (applicationFrameworkValue == EnabledValue)
+            {
+                if (unevaluatedPropertyValue.Contains("Form")) // Is there a better way to identify a form?
```

Comment:
I am a bit uneasy with this, but I am also not familiar with the project system specifics in the context.

What does `unevaluatedPropertyValue` contain here? If it is the Form's name, then this isn't going to work, since there is no guarantee that the string "Form" is part of the name. If it is the type, is this the fully qualified name?

If it's the name, do we have the means to do a type resolution based on the name without hopping to the WinForms Designer's server process to resolve the type?

Reply:
It depends. In Framework, the caption of the Startup _object_ changes to Startup _form_ if the AppFramework is enabled. And then only Forms are shown, rest is filtered out. 

My concern is that there was additional filtering code involved when getting back potential startup objects with the AppFramework enabled, and that those filters are missing here.

Here is a typical scenario in Framework without AppFramework enabled:

![image](https://user-images.githubusercontent.com/9663150/181343476-f7e97e39-8ae7-4bcd-be7f-7cc0328915b9.png)

And here, some project _with_ AppFramework enabled:

![image](https://user-images.githubusercontent.com/9663150/181343561-67b8fdfa-1f68-475a-81a7-0ac4560772b0.png)

Reply:
After discussing it offline, we've decided to remove the form validation and save the value in both the project file and the myapp file; I will revisit this soon after this.